internal/venice-common/src/main/java/com/linkedin/venice/utils/AvroRecordUtils.java

```diff
@@ -0,0 +1,235 @@
+      case STRING:
+      case BYTES:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case NULL:
+        return datum;
+      case RECORD:
+        return adaptRecordToSchema(expectedSchema, datum);
+      case ARRAY:
+        return adaptListToSchema(expectedSchema, datum);
+      case MAP:
+        return adaptMapToSchema(expectedSchema, datum);
+      case UNION:
+        return adaptUnionToSchema(expectedSchema, datum);
+      default:
+        // Defensive coding
+        throw new VeniceException("Unhandled Avro type. This is unexpected.");
+    }
+  }
+
+  private static Object adaptRecordToSchema(Schema expectedSchema, Object datum) {
+    if (!(datum instanceof IndexedRecord)) {
+      throw new VeniceException(); // TODO
+    }
+
+    IndexedRecord datumRecord = ((IndexedRecord) datum);
+    Schema datumSchema = datumRecord.getSchema();
+
+    if (datumSchema.equals(expectedSchema)) {
+      return datumRecord;
+    }
+
+    RecordDeserializer<IndexedRecord> deserializer =
+        FastSerializerDeserializerFactory.getAvroGenericDeserializer(datumSchema, expectedSchema);
+    RecordSerializer<IndexedRecord> serializer =
+        FastSerializerDeserializerFactory.getAvroGenericSerializer(datumSchema);
+
+    try {
+      return deserializer.deserialize(serializer.serialize(datumRecord));
+    } catch (VeniceSerializationException e) {
+      throw new VeniceException(e);
+    }
+  }
+
+  private static Collection<?> adaptListToSchema(Schema expectedSchema, Object datum) {
+    if (!(datum instanceof Collection)) {
+      throw new VeniceException(); // TODO
+    }
+
+    Collection<?> datumCollection = (Collection<?>) datum;
+
+    // Adapt List type - adapt all elements
+    Schema elementSchema = expectedSchema.getElementType();
+    if (AvroRecordUtils.canSchemaTypeBeAdapted(elementSchema.getType())) {
+      return datumCollection.stream()
+          .map(element -> AvroRecordUtils.adaptToSchema(elementSchema, element))
+          .collect(Collectors.toList());
+    } else {
+      return datumCollection;
+    }
+  }
+
+  private static Map<String, ?> adaptMapToSchema(Schema expectedSchema, Object datum) {
+    if (!(datum instanceof Map)) {
+      throw new VeniceException(); // TODO
+    }
+
+    // Adapt Map type - adapt all entries
+    Map<String, ?> datumMap = (Map<String, ?>) datum;
+    Schema elementSchema = expectedSchema.getValueType();
+    if (AvroRecordUtils.canSchemaTypeBeAdapted(elementSchema.getType())) {
+      return datumMap.entrySet()
+          .stream()
+          .collect(
+              Collectors
+                  .toMap(Map.Entry::getKey, entry -> AvroRecordUtils.adaptToSchema(elementSchema, entry.getValue())));
+    } else {
+      return datumMap;
+    }
+  }
+
+  private static Object adaptUnionToSchema(Schema expectedSchema, Object datum) {
+    int index;
+    try {
+      index = GenericData.get().resolveUnion(expectedSchema, datum);
```

Comment:
Can you check whether this function is available in the Venice supported Avro versions?
Basically 1.4 -> 1.10.

Reply:
`resolveUnion` is available in `1.4` as well. There was a difference in the `Exception` that is thrown.

The exception thrown in `1.4` (`AvroRuntimeException`) is different from `1.5+` (`UnresolvedUnionException`). Because of this, I had to `catch (Exception e)` instead of being specific.